### PR TITLE
Add missing MACA backend support for FP8, vector types, and lowering helpers

### LINKED ISSUE #27: MACA backend lacks dtype and code generation support required by FP8 and vectorised kernels
## Problem
The MACA backend is missing several pieces of dtype and code generation support required by the FP8 and vectorised example paths. As a result, some MACA examples fail before the example-level logic is even reached.

## Historical failures affected by this issue
- `examples/maca/cast/test_example_cast.py`
- `examples/maca/gemm_fp8/test_example_gemm_fp8.py`
- `examples/maca/deepseek_deepgemm/test_example_deepgemm_fp8_2xAcc.py`
- part of `examples/maca/gdn/test_example_gdn_compilation.py`
- part of `examples/maca/minference/test_vs_sparse_attn.py`
- part of `examples/maca/dequantize_gemm/test_example_dequantize_gemm.py`

## Technical causes
- TVM FFI cannot consume float8 tensors directly through the current DLPack path.
- MACA code generation lacks mappings for several float8 vector forms.
- MACA code generation cannot print or access some wide `float32` vector types used by the examples.
- MACA lowering is missing helper support used by shuffle-based or shared-region access paths.
- Dequantisation helpers need MACA-safe handling for unsigned packed data and exponent clamping.

## Minimal reproduction
- `python -m pytest -q examples/maca/cast/test_example_cast.py`
- `python -m pytest --runxfail -q examples/maca/gemm_fp8/test_example_gemm_fp8.py`

## Expected behaviour
The MACA backend should either lower these forms correctly or provide an explicit bridge layer so that the example code can compile and run without backend-specific failures.

### PR DESCRIPTION
**Pull request by VitalyAnkh**
Fixes #27.

## Problem
Several MACA example failures trace back to missing backend prerequisites rather than to the examples themselves. In particular, the MACA path was missing target-specific float8 runtime adaptation, the necessary MMA dtype metadata for the MACA tensor-core emitter, and a few lowering-side adjustments needed by the FP8 cast path.

## What this PR changes
- adds a MACA-specific TVM FFI adapter for float8 runtime handling, while keeping the generic adapter unchanged
- tightens the MACA codegen and lowering changes to the minimum target-specific surface required here
- folds the MACA MMA FP8 metadata into the existing dtype table so the emitter carries both storage and MMA input information
- keeps the backend changes focused on prerequisite support needed by later example ports
- adds regression coverage for the generic-vs-MACA adapter boundary and the MACA MMA dtype metadata

## Solution
This PR establishes the backend prerequisites required for the later MACA example adaptations, but it does not claim to clear the example xfail list on its own. The direct example-level fix in this branch is the MACA FP8 cast path, which now runs through the target-specific adapter correctly. Broader example portability and fallback work remains in the follow-up PRs.

## Alternatives considered
One option was to continue expanding shared runtime and DLPack code to accommodate MACA-specific float8 behavior. That would have pushed target-specific policy into generic components and made the boundary harder to reason about. Another option was to move all remediation into example code, but that would have left the underlying MACA backend prerequisites unresolved.

## Verification
- `python -m pytest -q examples/maca/cast/test_example_cast.py`
- `python -m pytest -q testing/python/jit/test_tilelang_jit_tvm_ffi.py -k 'dlpack_keeps_generic_float8_logical_dtype_mapping or generic_tvm_ffi_adapter_leaves_float8_tensor_untouched or maca_tvm_ffi_adapter_wraps_float8_for_runtime or jit_kernel_uses_target_specific_tvm_ffi_adapter'`
- `python -m pytest -q testing/python/kernel/test_tilelang_kernel_maca_mma_intrinsic.py`

## Stack context
This is the base backend-prerequisite PR in the MACA cleanup stack. The later PRs handle example-specific portability, launch-shape adjustments, and feature fallbacks.

### CODE REVIEW

File: src/target/codegen_maca.cc [hunk 1]
```diff
@@ -48,9 +48,12 @@ static std::string GetTileLangFP8Type(DataType type) {
...
            "for FP8";
   }
-  if (type.is_float8_e4m3() || type.is_float8_e4m3fn()) {
+  if (type.is_float8_e4m3() || type.is_float8_e4m3fn() ||
+      type.is_float8_e4m3fnuz() ||
+      type.code() == DataType::kFloat8_e4m3b11fnuz) {
     stream << "fp8_e4" << vec << "_t";
-  } else if (type.is_float8_e5m2()) {
+  } else if (type.is_float8_e5m2() || type.is_float8_e5m2fnuz() ||
+             type.code() == DataType::kFloat8_e5m2) {
```
Comment (Five-HZ) [inline src/target/codegen_maca.cc]: There is no need to add other types.

File: src/target/codegen_maca.cc [hunk 2]
```diff
@@ -362,6 +365,8 @@ void CodeGenTileLangMACA::PrintType(DataType t, std::ostream &os) { // NOLINT(*)
         ICHECK_EQ(lanes % 2, 0)
             << "only support even lane for float type with lanes > 4";
         os << "ulonglong" << lanes / 2;
+      } else if (lanes == 16 || lanes == 32) {
+        os << "float32x" << lanes;
```
Comment (Five-HZ) [inline src/target/codegen_maca.cc]: Using a large vector width resulted in the generation of x16 and x32 versions; should to restrict the width selection during vectorization.

File: tilelang/contrib/dlpack.py [hunk 3]
```diff
@@ -38,10 +38,10 @@ def convert_func(tvm_func, tensor_type, to_dlpack_func):
     import torch
 
     float8_dtype_map = {
-        torch.float8_e4m3fn: "float8_e4m3",
+        torch.float8_e4m3fn: "float8_e4m3fn",
         torch.float8_e4m3fnuz: "float8_e4m3fnuz",
         torch.float8_e5m2: "float8_e5m2",
-        torch.float8_e5m2fnuz: "float8_e5m2",
+        torch.float8_e5m2fnuz: "float8_e5m2fnuz",
```
Comment (Five-HZ) [inline tilelang/contrib/dlpack.py]: Keep the original mapping.

File: tilelang/intrinsics/maca_mma_macro_generator.py [hunk 4]
```diff
@@ -57,6 +57,14 @@ class TensorCoreIntrinEmitter:
...
     # Represent the thread binding in the form of (tx, warp_n, warp_m)
     is_m_first = False
+    fp8_dtypes = {
+        "float8_e4m3",
+        "float8_e5m2",
+        "float8_e4m3fn",
+        "float8_e5m2fn",
+        "float8_e4m3fnuz",
+        "float8_e5m2fnuz",
+    }
```
Comment (Five-HZ) [inline tilelang/intrinsics/maca_mma_macro_generator.py]: The type mappings should defined in `dtype_abbrv`.

File: tilelang/jit/adapter/tvm_ffi.py [hunk 5]
```diff
@@ -176,6 +176,27 @@ def _convert_torch_func(self) -> Callable[..., Any]:
...
+
+            float8_dtype = float8_dtype_map.get(arg.dtype)
+            if float8_dtype is None:
+                return arg
+
+            # tvm_ffi cannot ingest float8 tensors directly via DLPack today.
+            # Reuse the existing float8 bridge pattern: pass the storage as int8
+            # and recover the logical dtype through a TVM tensor view.
+            return runtime.from_dlpack(torch.utils.dlpack.to_dlpack(arg.view(torch.int8)))._create_view(arg.shape, dtype=float8_dtype)
+
```
Comment (Five-HZ) [inline tilelang/jit/adapter/tvm_ffi.py]: Support for the float8 type is already fully implemented; the adaptation for MACA is missing now, meaning no modifications to the core components are required.

File: src/target/codegen_maca.cc (hunk 1 above)
Comment (VitalyAnkh) [reply src/target/codegen_maca.cc]: Reduced this branch to the minimal MACA FP8 requirement. The extra `float8_e4m3b11fnuz` handling and the redundant `kFloat8_e5m2` branch are gone; only the fnuz variants used by the MACA examples remain.

File: tilelang/intrinsics/maca_mma_macro_generator.py (hunk 4 above)
Comment (VitalyAnkh) [reply tilelang/intrinsics/maca_mma_macro_generator.py]: The FP8 metadata is now folded into `dtype_abbrv`. Storage spelling, MMA suffix class, and the local MMA input dtype are derived from the same table instead of a separate side path.
